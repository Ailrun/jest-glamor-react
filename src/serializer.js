const isHtml = require('is-html')
const {replaceSelectors} = require('./replace-selectors')
const {
  getNodes,
  markNodes,
  getSelectors,
  getStylesAndAllSelectors,
} = require('./utils')

// doing this to make it easier for users to mock things
// like switching between development mode and whatnot.
const getGlamorStyleSheet = () => require('glamor').styleSheet

function createSerializer(styleSheet) {
  // eslint-disable-next-line complexity
  function test(val) {
    if (!val) {
      return false
    }
    return (
      isHtmlStringWithGlamorClasses(val) ||
      (!val.serializedWithJestGlamorReact &&
        (val.$$typeof === Symbol.for('react.test.json') ||
          (val instanceof HTMLElement && !isBeingSerialized(val))))
    )
  }

  function print(val, printer) {
    if (typeof val === 'string') {
      return fromHTMLString(val)
    }
    const nodes = getNodes(val)
    markNodes(nodes)
    const selectors = getSelectors(nodes)
    const {styles, allSelectors} = getStylesAndAllSelectors(
      selectors,
      styleSheet,
    )
    const printedVal = printer(val)
    // allows us to take multiple snapshots
    val.serializedWithJestGlamorReact = false
    if (styles) {
      return replaceSelectors(allSelectors, styles, printedVal)
    } else {
      return printedVal
    }
  }

  return {test, print}
}

function isBeingSerialized(node) {
  let currentNode = node

  while (currentNode) {
    if (currentNode.serializedWithJestGlamorReact) {
      return true
    }
    currentNode = currentNode.parentNode
  }
  return false
}

function isHtmlStringWithGlamorClasses(string) {
  return isHtml(string) && string.includes('css-')
}

function fromHTMLString(string) {
  const div = document.createElement('div')
  div.innerHTML = string
  const nodes = getNodes(div)
  const selectors = getSelectors(nodes)
  const {styles, allSelectors} = getStylesAndAllSelectors(
    selectors,
    getGlamorStyleSheet(),
  )
  return replaceSelectors(allSelectors, styles, string)
}

module.exports = Object.assign(createSerializer, {
  ...createSerializer(getGlamorStyleSheet),
  fromHTMLString,
})
