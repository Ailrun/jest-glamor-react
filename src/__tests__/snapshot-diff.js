import React from 'react'
import ReactDOM from 'react-dom'
import {Simulate} from 'react-dom/test-utils'
import * as glamor from 'glamor'
import {toMatchDiffSnapshot, getSnapshotDiffSerializer} from 'snapshot-diff'
import serializer, {fromHTMLString} from '../serializer'

expect.addSnapshotSerializer(getSnapshotDiffSerializer())
expect.addSnapshotSerializer(serializer)
expect.extend({toMatchDiffSnapshot})

function Button({count, ...props}) {
  const className = glamor.css({margin: 10 + count})
  return <button className={`${className}`} {...props} />
}

class Counter extends React.Component {
  state = {count: 0}
  increment = () => {
    this.setState(({count}) => ({count: count + 1}))
  }
  render() {
    const {count} = this.state
    return (
      <Button onClick={this.increment} count={count}>
        {count}
      </Button>
    )
  }
}

test('snapshot diff works', () => {
  const control = render(<Counter />)
  const variable = render(<Counter />)
  Simulate.click(variable)
  expect(fromHTMLString(control.outerHTML)).toMatchDiffSnapshot(
    fromHTMLString(variable.outerHTML),
  )
})

function render(ui) {
  const div = document.createElement('div')
  ReactDOM.render(ui, div)
  return div.children[0]
}

/* eslint react/prop-types:0 */
