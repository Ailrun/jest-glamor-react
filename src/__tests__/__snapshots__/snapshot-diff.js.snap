// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot diff works 1`] = `


Snapshot Diff:
- First value
+ Second value

  .css-0,
  [data-css-0] {
-   margin: 10px;
+   margin: 11px;
  }
  
- <button class="css-0">0</button>
+ <button class="css-0">1</button>
`;
